pkg/antctl/antctl.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2019 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package antctl
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/antctl/handlers"
+	"github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned/scheme"
+	"github.com/vmware-tanzu/antrea/pkg/version"
+)
+
+// unixDomainSockAddr is the address for antctl server in local mode.
+const unixDomainSockAddr = "/var/run/antctl.sock"
+
+var systemGroup = schema.GroupVersion{Group: "system.antrea.tanzu.vmware.com", Version: "v1beta1"}
+
+type transformedVersionResponse struct {
+	handlers.ComponentVersionResponse `json:",inline" yaml:",inline"`
+	AntctlVersion                     string `json:"antctlVersion" yaml:"antctlVersion"`
+}
+
+// versionTransform is the AddonTransform for the version command. This function
+// will try to parse the response as a ComponentVersionResponse and then populate
+// it with the version of antctl to a transformedVersionResponse object.
+func versionTransform(reader io.Reader, _ bool) (interface{}, error) {
+	b, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	klog.Infof("version transform received: %s", string(b))
+	cv := new(handlers.ComponentVersionResponse)
+	err = json.Unmarshal(b, cv)
+	if err != nil {
+		return nil, err
+	}
+	resp := &transformedVersionResponse{
+		ComponentVersionResponse: *cv,
+		AntctlVersion:            version.GetFullVersion(),
+	}
+	return resp, nil
+}
+
+// CommandList defines all commands that could be used in the antctl for both agent
+// and controller. The unit test "TestCommandListValidation" ensures it to be valid.
+var CommandList = &commandList{
+	definitions: []commandDefinition{
+		{
+			Use:                 "version",
+			Short:               "Print version information",
+			Long:                "Print version information of the antctl and the ${component}",
+			HandlerFactory:      new(handlers.Version),
+			GroupVersion:        &systemGroup,
+			TransformedResponse: new(transformedVersionResponse),
```

Comment:
If I understand the code correctly, this could be nil because we have an AddonTransform function?

Reply:
It should not be nil, this field is used to tell the output structure. If it is nil, the framework would not know how to format the result.
The workflow would look like:
```
(handler response struct) -> [default transform] -> (transformed response struct) -> [format]
                          \> [add-on transform] / 
```

Reply:
Understood, I was confused because of https://github.com/vmware-tanzu/antrea/pull/208/files#r366599836